dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,517 @@
+            )
+
+        if nones:
+            child_generator = iter(
+                filter(lambda child: child is not None, child_generator),
+            )  # pylint: disable=C0301
+
+        return child_generator
+
+    @property
+    def children(self) -> Iterator["Node"]:
+        """
+        returns an iterator of all nodes that are one step
+        from the current node down including through iterables
+        """
+        return self.fields(flat=True, nodes_only=True, obfuscated=False, nones=False)
+
+    def filter(self, func: Callable[["Node"], bool]) -> Iterator["Node"]:
+        """
+        find all nodes that `func` returns `True` for
+        """
+        if func(self):
+            yield self
+        for node in chain(*[child.filter(func) for child in self.children]):
+            yield node
+
+    def find_all(self, node_type: Type["Node"]) -> Iterator["Node"]:
+        """
+        find all nodes of a particular type in the node's sub-ast
+        """
+        return self.filter(lambda n: isinstance(n, node_type))
+
+    def apply(self, func: Callable[["Node"], None]):
+        """
+        traverse ast and apply func to each Node
+        """
+        func(self)
+        for child in self.children:
+            child.apply(func)
+
+    def compare(self, other: "Node") -> bool:
+        """
+        compare two ASTs
+        """
+
+        return self == other and all(
+            child.compare(other_child)
+            for child, other_child in zip_longest(self.children, other.children)
+        )
+
+    def __eq__(self, other) -> bool:
+        """
+        Compares two nodes for "top level" equality.
+        Checks for type equality and primitive field types for full equality.
+        Compares all others for type equality only. No recursing.
+        Note: Does not check (sub)AST. See `Node.compare` for comparing (sub)ASTs.
+        """
+        primitives = {int, float, str, bool, type(None)}
+        return type(self) == type(other) and all(  # pylint: disable=C0123
+            s == o
+            if type(s) in primitives  # pylint: disable=C0123
+            else type(s) == type(o)  # pylint: disable=C0123
+            for s, o in zip(
+                (self.fields(False, False, False, True)),
+                (other.fields(False, False, False, True)),
+            )
+        )
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        """
+        hash a node
+        """
+
+
+class Expression(Node):
+    """an expression type simply for type checking"""
+
+
+@dataclass(eq=False)  # type: ignore
+class Named(Expression):
+    """An Expression that has a name"""
+
+    name: str
+    quote_style: Optional[str]
+
+    @property
+    def quoted_name(self) -> str:
+        """
+        get the name of the Named Node including the quotes if any
+        """
+        return f'{self.quote_style if self.quote_style else ""}{self.name}{self.quote_style if self.quote_style else ""}'  # pylint: disable=C0301
```

Comment:
You can default `quote_style` to an empty string, then you can always use it:

```python
@dataclass(eq=False)  # type: ignore
class Named(Expression):
    """An Expression that has a name"""

    name: str
    quote_style: str = ""
```

I'd also create a `quoted_name` property to simplify the logic:

```python
@property
def quoted_name(self) -> str:
    return f"{self.quote_style}{self.name}{self.quote_style}"
```

@agorajek do we need to worry about asymetrical quoting, eg:

```
`foo'
```

Ie, quoting where the opening quote is different than the closing one.

Reply:
I'm not sure but I think we'd always need to catch this during parsing to raise an error since it might destroy the rest of the parse tree. A subsequent validation step might also work but that feels too late as I'm wondering if the parser would raise anyway just on some side-effect downstream.

Postgres raises something like: ```
ERROR: unterminated quoted string at or near `
```

Reply:
Made the default. The property exists already. As for the asymetrical quoting I did this little test with `sqloxide`:

```python
----> 1 parse_sql("select * from `a'", "ansi")

ValueError: Query parsing failed.
	sql parser error: Unterminated string literal at Line: 1, Column 17
```